File: dsp20.pas
(*********  PROGRAM NO.20    FILTER MAGNITUDE RESPONSE *********)
(***********  FROM Z-PLANE POLE AND ZERO LOCATIONS  ************)
program DSP20;
uses graph3;
var RP: array[1..20,1..2] of real;
var RZ: array[1..20,1..2] of real;
var CP: array[1..20,1..2] of real;
var CZ: array[1..20,1..2] of real;
var H: array[1..320] of real;
var A,B,W,C1,C2,S1,S2,D,R,T,E,MAX,DB,BASE,EX: real;
var NRP,NRZ,N,NCP,NCZ,K: integer;
label 1,2,3,4,5,6,7,8;
begin
    writeln('NO. OF SEPARATE REAL POLES'); readln(NRP);
      if NRP=0 then goto 1;
    writeln('ENTER VALUE, ORDER, OF EACH IN TURN');
    writeln('   FOLLOWED BY <RETURN>');
      for N:=1 to NRP do begin
        readln(RP[N,1],RP[N,2]); end;
 1: writeln('NO. OF SEPARATE REAL ZEROS'); readln(NRZ);
      if NRZ=0 then goto 2;
    writeln('ENTER VALUE, ORDER, OF EACH IN TURN');
    writeln('   FOLLOWED BY <RETURN>');
      for N:=1 to NRZ do begin
        readln(RZ[N,1],RZ[N,2]); end;
 2: writeln('NO. OF COMPLEX-CONJUGATE POLE PAIRS'); readln(NCP);
      if NCP=0 then goto 3;
    writeln('ENTER RADIUS, ANGLE (DEGREES), OF EACH');
    writeln('   IN TURN, FOLLOWED BY <RETURN>');
      for N:=1 to NCP do begin
        readln(CP[N,1],CP[N,2]); end;
 3: writeln('NO. OF COMPLEX-CONJUGATE ZERO PAIRS'); readln(NCZ);
      if NCZ=0 then goto 4;
    writeln('ENTER RADIUS, ANGLE (DEGREES), OF EACH');
    writeln('   IN TURN, FOLLOWED BY <RETURN>');
      for N:=1 to NCZ do begin
        readln(CZ[N,1],CZ[N,2]);end;
(************  COMPUTE FREQUENCY RESPONSE MAGNITUDE  ***********)
 4: for N:=1 to 320 do begin
      W:=3.141593*(N-1)/320;
      C1:=cos(W);C2:=cos(2*W);S1:=sin(W);S2:=sin(2*W);H[N]:=1;
        if NRP=0 then goto 5;
          for K:=1 to NRP do begin
            A:=RP[K,1];B:=RP[K,2];BASE:=1-2*A*C1+A*A;EX:=B/2;
            H[N]:=H[N]/(exp(EX*ln(BASE))); end;
 5:     if NRZ=0 then goto 6;
          for K:=1 to NRZ do begin
            A:=RZ[K,1];B:=RZ[K,2];BASE:=1-2*A*C1+A*A;EX:=B/2;
            H[N]:=H[N]*(exp(EX*ln(BASE))); end;
 6:      if NCP=0 then goto 7;
          for K:=1 to NCP do begin
            R:=CP[K,1];T:=CP[K,2]*3.141593/180;
            D:=C2-2*R*cos(T)*C1+R*R;E:=S2-2*R*cos(T)*S1;
            H[N]:=H[N]/(exp(0.5*ln(D*D+E*E))); end;
 7:      if NCZ=0 then goto 8;
          for K:=1 to NCZ do begin
            R:=CZ[K,1];T:=CZ[K,2]*3.141593/180;
            D:=C2-2*R*cos(T)*C1+R*R;E:=S2-2*R*cos(T)*S1;
            H[N]:=H[N]*(exp(0.5*ln(D*D+E*E))); end;
 8: end;
(*************  FIND MAXIMUM VALUE AND PRINT OUT  **************)
  graphmode;
  MAX:=0;
  for N:=1 to 320 do begin
    if abs(H[N])>MAX then MAX:=abs(H[N]); end;
  writeln('MAXIMUM GAIN=',20*ln(MAX)*0.4343,' dB');
(***************  DRAW RECTANGULAR GRID FOR PLOT  **************)
  for K:=0 to 5 do begin
   draw(0,30+20*K,320,30+20*K,1);
   draw(64*K,30,64*K,130,1);end;  draw(319,30,319,130,1);
(****  NORMALISE RESPONSE TO UNITY, CONVERT TO DB, AND PLOT  ***)
  for N:=1 to 320 do begin
    if abs(H[N])<0.000001 then H[N]:=0.000001;
    DB:=20*ln(abs(H[N])/MAX)*0.4343;if DB<-50 then DB:=-50;
    draw(N,130,N,round(30-2*DB),1); end;
end.
(***************************************************************)

File: dsp20.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define RPDIM     20
#define RZDIM     20
#define CPDIM     20
#define CZDIM     20
#define HDIM      320

float fH[HDIM + 1], fRP[RPDIM + 1][2], fRZ[RZDIM + 1][2],
      fCP[CPDIM + 1][2], fCZ[CZDIM + 1][2];

void main (void)
    {
    /*
     **** PROGRAM NO. 20  FILTER MAGNITUDE RESPONSE ************
     ****** FROM Z-PLANE POLE AND ZERO LOCATIONS ***************
     */
    int    iCount, iCount2, iNRP, iNRZ, iNCP, iNCZ;
    float  fW, fC1, fC2, fS1, fS2, fA, fB, fR, fT, fD, fE,
           fMAX, fDB;

    _clearscreen  (_GCLEARSCREEN);
    printf ("\n\nPROGRAM 20: Filter magnitude response from ");
    printf ("z-plane\npole and zero locations\n\n");

    printf ("Enter no. of SEPARATE real poles :");
    scanf ("%d", &iNRP);

    if (iNRP != 0)
        {
        printf ("\nEnter value, order of each in turn :\n");
        printf ("   Followed by <Return>.\n");

        for (iCount = 1; iCount <= iNRP; iCount++)
            scanf ("%f, %f", &fRP[iCount][0], &fRP[iCount][1]);
        }

    printf ("\nEnter no. of SEPARATE real zeros :");
    scanf ("%d", &iNRZ);

    if (iNRZ != 0)
        {
        printf ("\nEnter value, order of each in turn :\n");
        printf ("   Followed by <Return>.\n");

        for (iCount = 1; iCount <= iNRZ; iCount++)
            scanf ("%f, %f", &fRZ[iCount][0], &fRZ[iCount][1]);
        }

    printf ("\nEnter no. of complex-conjugate pole pairs :");
    scanf ("%d", &iNCP);

    if (iNCP != 0)
        {
        printf ("\nEnter radius, angle (in degrees) of ");
        printf ("each in turn :\n");
        printf ("   Followed by <Return>.\n");

        for (iCount = 1; iCount <= iNCP; iCount++)
            scanf ("%f, %f", &fCP[iCount][0], &fCP[iCount][1]);
        }

    printf ("\nEnter no. of complex-conjugate zero pairs :");
    scanf ("%d", &iNCZ);

    if (iNCZ != 0)
        {
        printf ("\nEnter radius, angle (in degrees) of ");
        printf ("each in turn :\n");
        printf ("   Followed by <Return>.\n");

        for (iCount = 1; iCount <= iNCZ; iCount++)
            scanf ("%f, %f", &fCZ[iCount][0], &fCZ[iCount][1]);
        }

    /*
     *********** COMPUTE FREQUENCY RESPONSE MAGNITUDE **********
     */
    for (iCount = 1; iCount <= HDIM; iCount++)
        {
        fW = 3.141593 * (float)(iCount - 1) / HDIM;
        fC1 = cos (fW);
        fC2 = cos (2.0 * fW);
        fS1 = sin (fW);
        fS2 = sin (2.0 * fW);
        fH[iCount] = 1.0;

        if (iNRP != 0)
            {
            for (iCount2 = 1; iCount2 <= iNRP; iCount2++)
                {
                fA = fRP[iCount2][0];
                fB = fRP[iCount2][1];
                fH[iCount] /= pow ((1.0 - 2.0 * fA * fC1 + fA * fA),
                                   (fB / 2.0));
                }
            }

        if (iNRZ != 0)
            {
            for (iCount2 = 1; iCount2 <= iNRZ; iCount2++)
                {
                fA = fRZ[iCount2][0];
                fB = fRZ[iCount2][1];
                fH[iCount] *= pow ((1.0 - 2.0 * fA * fC1 + fA * fA),
                                   (fB / 2.0));
                }
            }

        if (iNCP != 0)
            {
            for (iCount2 = 1; iCount2 <= iNCP; iCount2++)
                {
                fR = fCP[iCount2][0];
                fT = fCP[iCount2][1] * 3.141593 / 180.0;
                fD = fC2 - 2.0 * fR * cos (fT) * fC1 + fR * fR;
                fE = fS2 - 2.0 * fR * cos (fT) * fS1;

                fH[iCount] /= pow ((fD * fD + fE * fE), 0.5);
                }
            }

        if (iNCZ != 0)
            {
            for (iCount2 = 1; iCount2 <= iNCZ; iCount2++)
                {
                fR = fCZ[iCount2][0];
                fT = fCZ[iCount2][1] * 3.141593 / 180.0;
                fD = fC2 - 2.0 * fR * cos (fT) * fC1 + fR * fR;
                fE = fS2 - 2.0 * fR * cos (fT) * fS1;

                fH[iCount] *= pow ((fD * fD + fE * fE), 0.5);
                }
            }
        }

    /*
     ************** FIND MAXIMUM VALUE AND PRINT OUT ***********
     */
    fMAX = 0.0;

    for (iCount = 1; iCount <= HDIM; iCount++)
        if (fabs (fH[iCount] > fMAX))
            fMAX = fabs (fH[iCount]);

    printf ("\nMaximum gain = %f (%f DB)\n", fMAX,
            20.0 * log (fMAX) * 0.4343);

    printf ("Press any key to plot...");
    getch ();
    _setvideomode( _MRES16COLOR );

    /*
     ***************** DRAW RECTANGULAR GRID FOR PLOT **********
     */
    for (iCount = 0; iCount <= 5; iCount++)
        {
        _moveto (0, 30 + 20 * iCount);
        _lineto (320, 30 + 20 * iCount);
        _moveto (64 * iCount, 30);
        _lineto (64 * iCount, 130);
        }

    _moveto (319, 30);
    _lineto (319, 130);

    /*
     * NORMALISE RESPONSE TO UNITY, CONVERT TO DECIBELS & PLOT *
     */
    for (iCount = 1; iCount <= HDIM; iCount++)
        {
        if (fabs (fH[iCount]) < 0.000001)
            fH[iCount] = 0.000001;

        fDB = 20.0 * log (fabs (fH[iCount]) / fMAX) * 0.4343;

        if (fDB < -50.0)
            fDB = -50.0;

        _moveto (iCount, 130);
        _lineto (iCount, (int)(30.0 - 2.0 * fDB + 0.5));
        }

    getch();
    _setvideomode( _DEFAULTMODE );
    
    /*
     ***********************************************************
     */
    }
